lua/autorun/server/sv_cfc_logging_lib.moon

```diff
@@ -1,25 +1,57 @@
-export CFCAlertLib
+class RankManager
+    new: =>
+        -- Generate ordered ulx groups
+        flatten = (list, flattened) ->
+            flattened or= {}
+
+            for k, v in pairs list
+                table.insert flattened, k
+                return flatten v, flattened
+
+            flattened
+
+        ulx_groups = ULib.ucl.getInheritanceTree!
+
+        @ordered_groups = [rank for _,rank in ulx_groups]
+        @group_lookup = table.Reverse @ordered_groups
+
+    at_least: (rank) =>
+        min_index = @group_lookup[rank]
 
--- Generate ordered ulx groups
-flatten = (list, flattened) ->
-    flattened = flattened or {}
-    for k, v in pairs list
-        table.insert flattened, k
-        flatten v, flattened
+        [x for x=min_index, #@ordered_groups]
 
-    return flattened
+    at_most: (rank) =>
+        max_index = @group_lookup[rank]
 
-gScoreboard.OrderedGroups = {}
-ulxGroups = ULib.ucl.getInheritanceTree!
+        [@group_lookup[x] for x=0, max_index]
 
-for _, group in pairs flatten(ulxGroups)
-    table.insert gScoreboard.OrderedGroups, group
+    all: =>
+        @ordered_groups
 
 -- AlertingLibrary
+export CFCAlertLib
 class CFCAlertLib
+    new: (addon_name) =>
+        @addon_name = addon_name
+        @ranks = RankManager!
+
+    get_message: (message) =>
+        "[#{@addon_name}] #{message}}"
+
+    alert_ranks: (message, ranks) =>
+        rank_lookup = {v,true for _,v in pairs ranks}
+
+        for _, ply in pairs player.GetAll!
+            rank = ply\Team!
+            if rank_lookup[rank]
+                ply\ChatPrint @get_message(message)
+
+    alert_staff: (message) =>
+        ranks = @ranks\at_least("moderator")
```

Comment:
I was going to fight you on the variable name of `@ranks`, but it reads well using it like this.
It stays.

Reply:
;)